workflow-dockstore-metrics/workflow_dockstore_metrics/pull_and_send_metrics.py

```diff
@@ -0,0 +1,180 @@
+FROM workflow_runtime_info
+WHERE source_url IS NOT NULL
+ORDER BY source_url;
+'''
+
+class ExecutionData:
+    def __init__(self, query_row: bigquery.table.Row):
+        self.workflow_id = query_row.workflow_id
+        self.date_executed = query_row.workflow_start
+        # allowed statuses: "SUCCESSFUL", "FAILED", "ABORTED"
+        self.execution_status ="SUCCESSFUL"  if query_row.status == "Succeeded" else query_row.status.upper()
+        # represents duration in seconds in ISO 8601 duration format
+        self.execution_time = f"PT{query_row.workflow_runtime_time}S"
+
+
+class WorkflowData:
+
+    def __init__(self, query_row: bigquery.table.Row):
+        self.source_url = query_row.source_url
+        self.version = self.source_url.split('/')[5]
+
+        first_execution = ExecutionData(query_row)
+        self.workflow_executions = [first_execution]
+        self.sent_metric = False
+
+    def add_execution(self, execution: ExecutionData):
+        self.workflow_executions.append(execution)
+
+    def send_metrics(self, test: bool):
+
+        formatted_executions = ""
+        for execution in self.workflow_executions:
+            formatted_executions += f'''{{
+                "executionId": "{execution.workflow_id}",
+                "dateExecuted": "{execution.date_executed}",
+                "executionStatus": "{execution.execution_status}",
+                "executionTime":  "{execution.execution_time}",
+            }},
+            '''
+
+        json_request_body = f'''
+        {{
+            "runExecutions": [
+                {formatted_executions}
+            ],
+        }}
+        '''
+        uri = f"https://dockstore.org/api/api/ga4gh/v2/extended/{self.source_url}/versions/{self.version}/executions?platform=TERRA"
+        headers = {
+            "accept": "application/json",
+            "Content-Type": "application/json"
+        }
+
+        if test:
+            logging.info(f"[TEST MODE] Dockstore request: {json_request_body}")
+            status_code = 200
+        else:
+            response = requests.put(uri, json=json_request_body, headers=headers)
+            status_code = response.status_code
+
+        if status_code != 200:
+            logging.error(f"Error sending metrics for workflow {self.source_url} to dockstore for version {self.version}, response: {response.text}")
+        else:
+            logging.info(f"Successfully sent {len(self.workflow_executions)} execution metrics for workflow {self.source_url}, version {self.version} to dockstore")
+            self.sent_metric = True
+
+
+def update_metadata_table(client, workflow_ids: list[str], test: bool):
+    update_query = f'''
+      INSERT INTO `broad-dsde-prod-analytics-dev.warehouse.cromwell_metadata_sent_to_dockstore` (WORKFLOW_EXECUTION_ID) 
+       VALUES 
+       ('{"'), ('".join(workflow_ids)}');
```

Comment:
This was necessary instead of a new column because the cromwell_metadata table in BQ is actually a view and it seemed more of a hassle to update the associated tables by year than create a new reference table, but happy to discuss!

Reply:
I'm down to add another column!